_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. It is not permitted for one ticket to be standing while the other for the normal seated area for example.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class Seated(row: Int, column: Int) extends Seat
+case class Tribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />  
```

Comment:
can you put this image into `public` folder of the repo?

Reply:
I have uploaded the image, but left the code still pointing to wordpress. I only know how to link to the enear's blog repo after the blog post is deployed. (i will change it then).  
Or maybe there is some relative path I can use ?

Reply:
you can use `![type hierarchy](/public/imposing-invariants-in-scala-images/TypeHiearchy.png)`

browser adds domain to the url automatically